File: nodejs/tests/client.test.js
// 单测：客户端请求构建与信封解析（用本地 node:http 桩服务器，不依赖外网）。
import { test } from 'node:test';
import assert from 'node:assert/strict';
import http from 'node:http';

import { Client } from '../src/client.js';
import { Config, Environment } from '../src/config.js';
import { ApiError, TransportError } from '../src/errors.js';
import { sign } from '../src/signer.js';

const SECRET_PAY = 'sk_pay_0123456789abcdef0123456789abcdef';
const SECRET_PAYOUT = 'sk_payout_0123456789abcdef0123456789abc';

// 启动一个一次性桩服务器，回调收到的 (path, body)，并返回指定信封。
function startStub(handler) {
  return new Promise((resolveServer) => {
    const server = http.createServer((req, res) => {
      const chunks = [];
      req.on('data', (c) => chunks.push(c));
      req.on('end', () => {
        const body = JSON.parse(Buffer.concat(chunks).toString('utf8'));
        const envelope = handler(req.url, body, res);
        if (res.writableEnded) return;
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(envelope ?? { code: 0, message: 'ok', data: {} }));
      });
    });
    server.listen(0, '127.0.0.1', () => {
      const { port } = server.address();
      resolveServer({ server, baseUrl: `http://127.0.0.1:${port}/api/open/v1` });
    });
  });
}

function makeClient(baseUrl) {
  return new Client(
    new Config({
      merchantNo: 'M00000001',
      apiKey: 'ak_demo_key',
      apiSecretPay: SECRET_PAY,
      apiSecretPayout: SECRET_PAYOUT,
      baseUrl,
    }),
  );
}

test('Environment 预设基址正确', () => {
  assert.equal(Environment.PRODUCTION, 'https://api.project-p-merchant.cniia.cloud/api/open/v1');
  assert.equal(Environment.SANDBOX, 'http://127.0.0.1:3090/api/open/v1');
});

test('Config baseUrl 覆盖优先于 environment 且去除末尾斜杠', () => {
  const c = new Config({
    merchantNo: 'M',
    apiKey: 'k',
    environment: Environment.PRODUCTION,
    baseUrl: 'https://api.custom.example/api/open/v1/',
  });
  assert.equal(c.baseUrl, 'https://api.custom.example/api/open/v1');
});

test('payCreate：注入通用字段/timestamp/nonce，过滤 null，签名用 pay 密钥，命中正确 path', async () => {
  let captured = null;
  const { server, baseUrl } = await startStub((url, body) => {
    captured = { url, body };
    return { code: 0, message: 'ok', data: { order_no: 'P1', status: 'pending' } };
  });
  try {
    const client = makeClient(baseUrl);
    const { data, raw } = await client.payCreate({
      out_order_no: '202501010001',
      amount: 10000,
      currency: 'PHP',
      pay_method: 'gcash',
      country: 'PH',
      notify_url: 'https://merchant.example.com/api/notify/pay',
      return_url: null, // 应被过滤
    });

    assert.equal(captured.url, '/api/open/v1/merchant/pay/create');
    // 通用字段注入
    assert.equal(captured.body.merchant_no, 'M00000001');
    assert.equal(captured.body.api_key, 'ak_demo_key');
    assert.equal(typeof captured.body.timestamp, 'number');
    assert.equal(typeof captured.body.nonce, 'string');
    assert.ok(captured.body.nonce.length > 0);
    // null 字段被过滤
    assert.ok(!('return_url' in captured.body));
    // 签名用 pay 密钥，且对收到的（去 sign）体可复现
    const { sign: gotSign, ...rest } = captured.body;
    assert.equal(gotSign, sign(rest, SECRET_PAY));
    // 信封解析
    assert.equal(data.order_no, 'P1');
    assert.equal(raw.code, 0);
  } finally {
    server.close();
  }
});

test('每请求 nonce 唯一', async () => {
  const seen = [];
  const { server, baseUrl } = await startStub((url, body) => {
    seen.push(body.nonce);
    return { code: 0, message: 'ok', data: {} };
  });
  try {
    const client = makeClient(baseUrl);
    await client.balanceQuery({});
    await client.balanceQuery({});
    assert.equal(seen.length, 2);
    assert.notEqual(seen[0], seen[1]);
  } finally {
    server.close();
  }
});

test('payoutCreate：签名用 payout 密钥', async () => {
  let captured = null;
  const { server, baseUrl } = await startStub((url, body) => {
    captured = body;
    return { code: 0, message: 'ok', data: { payout_no: 'W1', status: 'pending' } };
  });
  try {
    const client = makeClient(baseUrl);
    await client.payoutCreate({
      out_payout_no: 'WD1',
      amount: 100000,
      currency: 'PHP',
      pay_method: 'bank',
      country: 'PH',
      notify_url: 'https://m.example.com/cb',
      account_no: '1234567890',
      account_name: 'San Zhang',
      bank_code: 'BDO',
    });
    const { sign: gotSign, ...rest } = captured;
    assert.equal(gotSign, sign(rest, SECRET_PAYOUT));
    // 用 pay 密钥应当对不上
    assert.notEqual(gotSign, sign(rest, SECRET_PAY));
  } finally {
    server.close();
  }
});

test('payoutReceiptQuery：inline 以整数 1/0 发送', async () => {
  const bodies = [];
  const { server, baseUrl } = await startStub((url, body) => {
    bodies.push(body);
    return { code: 0, message: 'ok', data: {} };
  });
  try {
    const client = makeClient(baseUrl);
    await client.payoutReceiptQuery({ payout_no: 'W1', inline: true });
    await client.payoutReceiptQuery({ payout_no: 'W1', inline: false });
    await client.payoutReceiptQuery({ payout_no: 'W1' }); // 省略
    assert.equal(bodies[0].inline, 1);
    assert.strictEqual(bodies[0].inline === true, false); // 不是布尔
    assert.equal(bodies[1].inline, 0);
    assert.ok(!('inline' in bodies[2]));
  } finally {
    server.close();
  }
});

test('code !== 0 抛 ApiError，携带 code/message/data/raw', async () => {
  const { server, baseUrl } = await startStub(() => ({
    code: 100104,
    message: '签名错误',
    data: { missing_fields: ['x'] },
  }));
  try {
    const client = makeClient(baseUrl);
    await assert.rejects(
      () => client.payQuery({ out_order_no: 'x' }),
      (err) => {
        assert.ok(err instanceof ApiError);
        assert.equal(err.code, 100104);
        assert.equal(err.message, '签名错误');
        assert.deepEqual(err.data, { missing_fields: ['x'] });
        assert.equal(err.raw.code, 100104);
        return true;
      },
    );
  } finally {
    server.close();
  }
});

test('HTTP 非 2xx 抛 TransportError', async () => {
  const { server, baseUrl } = await startStub((url, body, res) => {
    res.writeHead(500, { 'Content-Type': 'text/plain' });
    res.end('boom');
  });
  try {
    const client = makeClient(baseUrl);
    await assert.rejects(
      () => client.balanceQuery({}),
      (err) => {
        assert.ok(err instanceof TransportError);
        assert.equal(err.statusCode, 500);
        assert.equal(err.body, 'boom');
        return true;
      },
    );
  } finally {
    server.close();
  }
});

test('响应非合法 JSON 抛 TransportError', async () => {
  const { server, baseUrl } = await startStub((url, body, res) => {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end('not-json');
  });
  try {
    const client = makeClient(baseUrl);
    await assert.rejects(() => client.payMethodsQuery({}), TransportError);
  } finally {
    server.close();
  }
});

test('全部 11 个端点方法均存在', () => {
  const client = makeClient('http://127.0.0.1:1/api/open/v1');
  const methods = [
    'payCreate',
    'payQuery',
    'payMethodsQuery',
    'balanceQuery',
    'payTestComplete',
    'payoutCreate',
    'payoutQuery',
    'payoutBanksQuery',
    'payoutProofQuery',
    'payoutReceiptQuery',
    'payoutTestComplete',
  ];
  for (const m of methods) {
    assert.equal(typeof client[m], 'function', `missing method ${m}`);
  }
});

test('调用 pay 类端点但未配 apiSecretPay 抛 TypeError', async () => {
  const client = new Client(
    new Config({ merchantNo: 'M', apiKey: 'k', apiSecretPayout: SECRET_PAYOUT, baseUrl: 'http://127.0.0.1:1/x' }),
  );
  await assert.rejects(() => client.payCreate({ out_order_no: 'x', amount: 1 }), TypeError);
});

File: go/client_test.go
package projectp

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

// TestEnvironmentBaseURL 校验预设基址与自定义覆盖。
func TestEnvironmentBaseURL(t *testing.T) {
	if got := Production.BaseURL(); got != productionBaseURL {
		t.Errorf("production 基址错误: %s", got)
	}
	if got := Sandbox.BaseURL(); got != sandboxBaseURL {
		t.Errorf("sandbox 基址错误: %s", got)
	}

	c := NewClient(Config{Environment: Sandbox})
	if c.baseURL != sandboxBaseURL {
		t.Errorf("sandbox client 基址错误: %s", c.baseURL)
	}

	custom := "https://api.agent.example.com/api/open/v1"
	c2 := NewClient(Config{Environment: Production, BaseURL: custom})
	if c2.baseURL != custom {
		t.Errorf("自定义基址未覆盖预设: %s", c2.baseURL)
	}
}

// TestBuildBodyInjectsCommonFields 校验通用字段、nonce、timestamp、sign 注入与 nil 过滤。
func TestBuildBodyInjectsCommonFields(t *testing.T) {
	// 固定时钟与 nonce 便于断言。
	origNow, origNonce := nowUnix, newNonce
	nowUnix = func() int64 { return 1736073600 }
	newNonce = func() string { return "fixed_nonce" }
	defer func() { nowUnix, newNonce = origNow, origNonce }()

	c := NewClient(Config{
		MerchantNo: "M00000001",
		APIKey:     "ak_demo_key",
		SecretPay:  "sk_test_0123456789abcdef0123456789abcdef",
	})

	body := c.buildBody(map[string]any{
		"out_order_no": "202501010001",
		"amount":       10000,
		"return_url":   nil, // 应被过滤。
	}, c.cfg.SecretPay)

	if body["merchant_no"] != "M00000001" || body["api_key"] != "ak_demo_key" {
		t.Errorf("通用字段缺失: %+v", body)
	}
	if body["timestamp"] != int64(1736073600) {
		t.Errorf("timestamp 未注入: %+v", body["timestamp"])
	}
	if body["nonce"] != "fixed_nonce" {
		t.Errorf("nonce 未注入: %+v", body["nonce"])
	}
	if _, ok := body["return_url"]; ok {
		t.Errorf("nil 字段未被过滤")
	}
	sign, ok := body["sign"].(string)
	if !ok || len(sign) != 64 {
		t.Errorf("sign 未正确注入: %+v", body["sign"])
	}
	// sign 应等于对去掉 sign 后的 body 重新计算的值（自洽）。
	verifyBody := cloneParams(body)
	if !VerifyCallback(verifyBody, c.cfg.SecretPay) {
		t.Errorf("注入的 sign 与重算不一致")
	}
}

// TestCallSecretSelection 校验 pay 用 SecretPay、payout 用 SecretPayout。
func TestCallSecretSelection(t *testing.T) {
	c := NewClient(Config{SecretPay: "pay_secret", SecretPayout: "payout_secret"})
	if c.secretFor(usePay) != "pay_secret" {
		t.Errorf("pay 密钥选择错误")
	}
	if c.secretFor(usePayout) != "payout_secret" {
		t.Errorf("payout 密钥选择错误")
	}
}

// TestReceiptInlineAsInteger 校验 inline 以整数 1/0 发送。
func TestReceiptInlineAsInteger(t *testing.T) {
	var captured map[string]any
	srv := newTestServer(t, &captured, `{"code":0,"message":"ok","data":{"receipt_url":"u"}}`)
	defer srv.Close()

	c := NewClient(Config{BaseURL: srv.URL, SecretPayout: "s"})

	if _, err := c.PayoutReceiptQuery(context.Background(), map[string]any{"payout_no": "W1"}, true); err != nil {
		t.Fatalf("inline=true 调用失败: %v", err)
	}
	if n, ok := captured["inline"].(json.Number); !ok || n.String() != "1" {
		t.Errorf("inline=true 应以整数 1 发送，实际: %+v (%T)", captured["inline"], captured["inline"])
	}

	if _, err := c.PayoutReceiptQuery(context.Background(), map[string]any{"payout_no": "W1"}, false); err != nil {
		t.Fatalf("inline=false 调用失败: %v", err)
	}
	if n, ok := captured["inline"].(json.Number); !ok || n.String() != "0" {
		t.Errorf("inline=false 应以整数 0 发送，实际: %+v (%T)", captured["inline"], captured["inline"])
	}
}

// TestCallBusinessError 校验 code!=0 抛 APIError 并携带 code/message/data。
func TestCallBusinessError(t *testing.T) {
	var captured map[string]any
	srv := newTestServer(t, &captured,
		`{"code":100001,"message":"\"amount\" must be a number","data":{"missing_fields":["amount"]}}`)
	defer srv.Close()

	c := NewClient(Config{BaseURL: srv.URL, SecretPay: "s"})
	resp, err := c.PayCreate(context.Background(), map[string]any{"out_order_no": "x"})
	if err == nil {
		t.Fatalf("应返回业务错误")
	}
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("应为 *APIError，实际 %T", err)
	}
	if apiErr.Code != 100001 {
		t.Errorf("code 错误: %d", apiErr.Code)
	}
	if apiErr.Data["missing_fields"] == nil {
		t.Errorf("data 未携带")
	}
	// 即便出错，也应能拿到原始响应。
	if resp == nil || len(resp.Raw) == 0 {
		t.Errorf("应保留原始响应途径")
	}
}

// TestCallTransportError 校验非 2xx 抛 TransportError。
func TestCallTransportError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadGateway)
		_, _ = io.WriteString(w, "upstream down")
	}))
	defer srv.Close()

	c := NewClient(Config{BaseURL: srv.URL, SecretPay: "s", Timeout: 5 * time.Second})
	_, err := c.PayQuery(context.Background(), map[string]any{"order_no": "x"})
	if err == nil {
		t.Fatalf("应返回传输错误")
	}
	tErr, ok := err.(*TransportError)
	if !ok {
		t.Fatalf("应为 *TransportError，实际 %T", err)
	}
	if tErr.StatusCode != http.StatusBadGateway {
		t.Errorf("状态码错误: %d", tErr.StatusCode)
	}
}

// TestCallSuccessParsesData 校验成功响应解析 data 且大整数不被科学计数法污染。
func TestCallSuccessParsesData(t *testing.T) {
	var captured map[string]any
	srv := newTestServer(t, &captured,
		`{"code":0,"message":"ok","data":{"order_no":"P1","amount":1000000000000,"status":"pending"}}`)
	defer srv.Close()

	c := NewClient(Config{BaseURL: srv.URL, SecretPay: "s"})
	resp, err := c.PayCreate(context.Background(), map[string]any{"out_order_no": "x", "amount": 1000000000000})
	if err != nil {
		t.Fatalf("调用失败: %v", err)
	}
	amt, ok := resp.Data["amount"].(json.Number)
	if !ok || amt.String() != "1000000000000" {
		t.Errorf("大整数解析错误: %+v (%T)", resp.Data["amount"], resp.Data["amount"])
	}
}

// newTestServer 返回一个回放固定响应、并把请求体捕获到 captured 的测试服务器。
func newTestServer(t *testing.T, captured *map[string]any, response string) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		dec := json.NewDecoder(bytes.NewReader(raw))
		dec.UseNumber()
		var m map[string]any
		_ = dec.Decode(&m)
		*captured = m
		w.Header().Set("Content-Type", "application/json")
		_, _ = io.WriteString(w, response)
	}))
}
